druid/src/main/java/org/apache/calcite/adapter/druid/DruidConnectionImpl.java

```diff
@@ -338,8 +338,40 @@ private void parseField(List<String> fieldNames, List<ColumnMetaData.Rep> fieldT
       break;
     case VALUE_STRING:
     default:
-      rowBuilder.set(i, parser.getText());
-      break;
+      String s = parser.getText();
+      if (type != null) {
+        switch (type) {
+        case LONG:
+        case PRIMITIVE_LONG:
+        case SHORT:
+        case PRIMITIVE_SHORT:
+        case INTEGER:
+        case PRIMITIVE_INT:
+          if (s.equals("Infinity") || s.equals("-Infinity") || s.equals("NaN")) {
+            throw new RuntimeException("/ by zero");
+          }
+        case FLOAT:
+        case PRIMITIVE_FLOAT:
+        case PRIMITIVE_DOUBLE:
+        case NUMBER:
+        case DOUBLE:
+          if (s.equals("Infinity")) {
+            rowBuilder.set(i, Double.POSITIVE_INFINITY);
+            break;
```

Comment:
I am not sure what do you mean. Do you prefer:
```java
case FLOAT:
  //fallthrough
case PRIMITIVE_FLOAT:
  //fallthrough
case PRIMITIVE_DOUBLE:
  //fallthrough
case NUMBER:
  //fallthrough
case DOUBLE:
          if (s.equals("Infinity")) {
            rowBuilder.set(i, Double.POSITIVE_INFINITY);
            break;
          } else if (s.equals("-Infinity")) {
            rowBuilder.set(i, Double.NEGATIVE_INFINITY);
            break;
          } else if (s.equals("NaN")) {
            rowBuilder.set(i, Double.NaN);
            break;
          }
        //fallthrough
        default:
```
or just:
```java
  case FLOAT:
  case PRIMITIVE_FLOAT:
  case PRIMITIVE_DOUBLE:
  case NUMBER:
  case DOUBLE:
          if (s.equals("Infinity")) {
            rowBuilder.set(i, Double.POSITIVE_INFINITY);
            break;
          } else if (s.equals("-Infinity")) {
            rowBuilder.set(i, Double.NEGATIVE_INFINITY);
            break;
          } else if (s.equals("NaN")) {
            rowBuilder.set(i, Double.NaN);
            break;
          }
         //fallthrough
        default:
```
I think the first one is what you mean, but it seems a little bit cumbersome. The second is good enough to clarify the confusion in here.